File: LeetCode/1600. Throne Inheritance/solution.py
class ThroneInheritance:

    def __init__(self, kingName: str):
        self.king = kingName
        self.alive = set([kingName])
        self.children = defaultdict(list)


    def birth(self, parentName: str, childName: str) -> None:
        self.children[parentName].append(childName)
        self.alive.add(parentName)
        self.alive.add(childName)


    def death(self, name: str) -> None:
        self.alive.remove(name)


    def getInheritanceOrder(self) -> List[str]:
        curOrder = []
        def dfs(x):
            if x in self.alive:
                curOrder.append(x)
            for nxt in self.children[x]:
                dfs(nxt)
            return
        dfs(self.king)
        return curOrder






# Your ThroneInheritance object will be instantiated and called as such:
# obj = ThroneInheritance(kingName)
# obj.birth(parentName,childName)
# obj.death(name)
# param_3 = obj.getInheritanceOrder()

File: LeetCode/1600. Throne Inheritance/solution.cpp
class ThroneInheritance {
    string root;
    unordered_map<string,vector<string>> children;
    unordered_set<string> dead;

    void dfs(string cur, vector<string> &order){
        if (dead.find(cur) == dead.end())
            order.push_back(cur);
        for (string &nxt : children[cur]){
            dfs(nxt, order);
        }
    }
public:
    ThroneInheritance(string kingName) {
        root = kingName;
    }
    
    void birth(string parentName, string childName) {
        children[parentName].push_back(childName);
    }
    
    void death(string name) {
        dead.insert(name);
    }
    
    vector<string> getInheritanceOrder() {
        vector<string> order;
        dfs(root, order);
        return order;
    }
};

/**
 * Your ThroneInheritance object will be instantiated and called as such:
 * ThroneInheritance* obj = new ThroneInheritance(kingName);
 * obj->birth(parentName,childName);
 * obj->death(name);
 * vector<string> param_3 = obj->getInheritanceOrder();
 */